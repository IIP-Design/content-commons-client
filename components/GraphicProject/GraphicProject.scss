@import "styles/colors.scss";

.graphic-project {

  &__popover {

    .popover_content {
      top: calc( 100% + 8px );

      &:before {
        top: -8px;
        width: 15px;
        height: 8px;
        margin-right: 2px;
        background: $bright-blue;        
      }

      &:after {
        content: "";
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        padding-top: 5px;
        background-image: linear-gradient(90deg, $light-blue 3%, $bright-blue 100%);
      }
    }

    &--share {
      margin-right: 5px;
      .popover_content {
        font-size: 1rem;        
      }
    }
  }
}