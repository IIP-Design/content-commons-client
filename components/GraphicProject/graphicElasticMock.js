export const graphicElasticMock = [
  {
    _source: {
      id: 'ck8h90w6x030l0765onryiqdt',
      site: 'commons.america.gov',
      type: 'graphic',
      title: 'Graphic Elastic Mock Data',
      // projectType: 'Social Media Graphic',
      published: '2020-03-22T15:15:00.074Z',
      modified: '2020-04-02T15:15:00.074Z',
      visibility: 'PUBLIC',
      owner: 'GPA Design & Editorial',
      // alt: 'Some alt text here...',
      descPublic: {
        visibility: 'PUBLIC',
        content:
                 "Secretary Pompeo outlined 12 requirements for Iran's regime to act like a normal state. With the November 4th sanctions deadline, here are reminders for the regime, starting with #12: Declare prior military dimensions of your nuclear program & abandon such work.",
      },
      descInternal: {
        visibility: 'INTERNAL',
        content: 'Internal description....',
      },
      copyright: 'COPYRIGHT',
      images: [
        {
          id: 'eniebfwkjfenwknfwqjn',
          title: 'Mon titre graphique',
          visibility: 'PUBLIC',
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_FRENCH_Twitter_Output.png',
          signedUrl: 'https://mock-signed.png',
          language: {
            language_code: 'fr',
            text_direction: 'ltr',
            locale: 'fr-fr',
            display_name: 'French',
            native_name: 'French',
          },
          filename: 'CAPTIONS_Made_in_America_FRENCH_Twitter_Output.png',
          alt: 'Some alt text here...',
          filesize: '600 MB',
          width: 800,
          height: 600,
          social: ['Twitter'],
          style: 'Quote',
        },
        {
          id: 'vewjnewjnewnwqfjn',
          title: 'Mon titre graphique FACEBOOK',
          visibility: 'PUBLIC',
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_FRENCH_Facebook_Output.png',
          signedUrl: 'https://mock-signed.png',
          language: {
            language_code: 'fr',
            text_direction: 'ltr',
            locale: 'fr-fr',
            display_name: 'French',
            native_name: 'French',
          },
          filename: 'CAPTIONS_Made_in_America_FRENCH_Facebook_Output.png',
          alt: 'Some alt text here...',
          filesize: '600 MB',
          width: 800,
          height: 600,
          social: ['Facebook', 'Instagram'],
          style: 'Quote',
        },
        {
          id: 'dqwkq4dm9qwlknlkwfn',
          title: 'My graphic title',
          visibility: 'PUBLIC',
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_English_Output.png',
          signedUrl: 'https://mock-signed.png',
          language: {
            language_code: 'en',
            text_direction: 'ltr',
            locale: 'en-us',
            display_name: 'English',
            native_name: 'English',
          },
          filename: 'CAPTIONS_Made_in_America_English_Output.png',
          alt: 'Some alt text here...',
          filesize: '600 MB',
          width: 800,
          height: 600,
          social: ['Twitter'],
          style: 'Quote',
        },
      ],
      supportFiles: [
        {
          id: 'fniuf2iuh3uh123iuh23uh',
          visibility: 'INTERNAL',
          editable: true,
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_English_Output.psd',
          language: {
            language_code: 'en',
            text_direction: 'ltr',
            locale: 'en-us',
            display_name: 'English',
            native_name: 'English',
          },
        },
        {
          id: 'fnwndf2j3nlkf3n23uh',
          visibility: 'INTERNAL',
          editable: true,
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_English_Output.ai',
          language: {
            language_code: 'en',
            text_direction: 'ltr',
            locale: 'en-us',
            display_name: 'English',
            native_name: 'English',
          },
        },
        {
          id: '32kbf32kjbk2f3jbf3j2',
          visibility: 'INTERNAL',
          editable: true,
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_English_Output.png',
          language: {
            language_code: 'en',
            text_direction: 'ltr',
            locale: 'en-us',
            display_name: 'English',
            native_name: 'English',
          },
        },
        {
          id: 'wnljnewln299jnjkwnewef',
          visibility: 'INTERNAL',
          editable: true,
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_French_Output.psd',
          language: {
            language_code: 'fr',
            text_direction: 'ltr',
            locale: 'fr-fr',
            display_name: 'French',
            native_name: 'French',
          },
        },
        {
          id: 'qewbr438h4398h43bu3feiu',
          visibility: 'INTERNAL',
          editable: true,
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_French_Output.ai',
          language: {
            language_code: 'fr',
            text_direction: 'ltr',
            locale: 'fr-fr',
            display_name: 'French',
            native_name: 'French',
          },
        },
        {
          id: 'beqwibfu4389h34hh34h43uf',
          visibility: 'INTERNAL',
          editable: false,
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_French_Output.doc',
          language: {
            language_code: 'fr',
            text_direction: 'ltr',
            locale: 'fr-fr',
            display_name: 'French',
            native_name: 'French',
          },
        },
        {
          id: 'fnu2h348hf4842h8',
          visibility: 'INTERNAL',
          editable: false,
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_French_Output.pdf',
          language: {
            language_code: 'fr',
            text_direction: 'ltr',
            locale: 'fr-fr',
            display_name: 'French',
            native_name: 'French',
          },
        },
        {
          id: 'f23u4h824f2ewkjnfkjwfnkjn',
          visibility: 'INTERNAL',
          editable: false,
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_English_Output.doc',
          language: {
            language_code: 'en',
            text_direction: 'ltr',
            locale: 'en-us',
            display_name: 'English',
            native_name: 'English',
          },
        },
        {
          id: '2323uh32uhr328223hfuhb',
          visibility: 'INTERNAL',
          editable: false,
          url:
                   'https://cdp-video-tst.s3.amazonaws.com/2020/04/commons.america.gov_ck8h90w6x030l0765onryiqdt/CAPTIONS_Made_in_America_English_Output.pdf',
          language: {
            language_code: 'en',
            text_direction: 'ltr',
            locale: 'en-us',
            display_name: 'English',
            native_name: 'English',
          },
        },
      ],
      tags: [
        {
          name: 'biomedical science',
          id: 'KBb5K3EB-L8EaltkVCV8',
        },
      ],
      categories: [
        {
          name: 'science & technology',
          id: 'TBb5K3EB-L8EaltkViXQ',
        },
      ],
    },
  },
];
