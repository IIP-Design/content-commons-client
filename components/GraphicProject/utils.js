import { getTransformedLangTaxArray, getCount } from 'lib/utils';

const structureLangObj = item => ( {
  language: {
    language_code: item.language.languageCode,
    text_direction: item.language.textDirection,
    locale: item.language.locale,
    display_name: item.language.displayName,
    native_name: item.language.nativeName,
  },
} );

export const normalizeGraphicProjectByAPI = ( { file, useGraphQl = false } ) => {
  const graphicObj = {
    id: file.id || '',
    type: 'graphic',
    site: file.site || '',
    title: file.title || '',
    projectType: file.projectType || '',
    published: file.published || '',
    modified: file.modified || '',
    owner: file.owner || '',
    desc: file.desc || '',
    descInternal: file.descInternal || '',
    copyright: file.copyright || '',
    images: file.images || [],
    supportFiles: file.supportFiles || [],
    categories: file.categories || '',
  };

  // Elastic API
  if ( !useGraphQl ) {
    const imagesWithFileNameProp = file.images.map( img => ( {
      ...img,
      filename: img.filename,
    } ) );

    const supportFilesWithFileNameProp = Array.isArray( file.supportFiles ) && file.supportFiles.map( supFile => {
      const filenameFromSrc = supFile.url.slice( supFile.url.lastIndexOf( '/' ) + 1 );

      return {
        ...supFile,
        filename: filenameFromSrc,
      };
    } );

    const esObj = {
      images: imagesWithFileNameProp || [],
      supportFiles: supportFilesWithFileNameProp || [],
    };

    return { ...graphicObj, ...esObj };
  }

  // GraphQL API
  if ( useGraphQl ) {
    const structuredImages = file.images.map( img => {
      const { dimensions } = img;

      return {
        ...img,
        ...structureLangObj( img ),
        url: img.signedUrl,
        width: dimensions.width,
        height: dimensions.height,
      };
    } );

    const structuredSupportFiles = Array.isArray( file.supportFiles ) && file.supportFiles.map( supFile => ( {
      ...supFile,
      ...structureLangObj( supFile ),
      srcUrl: supFile.signedUrl,
    } ) );

    const gqlObj = {
      alt: file.alt || '',
      projectType: file.type || '',
      published: file.publishedAt || '',
      modified: file.updatedAt || '',
      owner: file.team?.name || '',
      desc: file.descPublic || '',
      images: structuredImages || [],
      supportFiles: structuredSupportFiles || [],
      categories: getTransformedLangTaxArray( file.categories ) || [],
    };

    return { ...graphicObj, ...gqlObj };
  }

  return graphicObj;
};

// Use Twitter graphics as default for display otherwise whatever graphic image is available
export const filterGraphicImgs = images => {
  const containsTwitterImgs = images.some( img => img.social.includes( 'Twitter' ) );

  if ( containsTwitterImgs ) {
    return images.filter( img => img.social.includes( 'Twitter' ) );
  }

  return images;
};

export const getGraphicImgsBySocial = ( images, platform, useGraphQl = false ) => {
  if ( !getCount( images ) ) return [];

  const filteredImgs = images.reduce( ( allImgs, img ) => {
    let platformImgs = [];
    let condition = img.social.includes( platform );

    if ( useGraphQl ) {
      platformImgs = img.social.filter( s => s.name === platform );
      condition = getCount( platformImgs );
    }

    if ( condition ) {
      allImgs.push( img );
    }

    return allImgs;
  }, [] );

  return getCount( filteredImgs ) ? filteredImgs : images;
};
