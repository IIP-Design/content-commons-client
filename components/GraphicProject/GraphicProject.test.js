import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { act } from 'react-dom/test-utils';
import GraphicProject from './GraphicProject';
import { getTransformedLangTaxArray } from 'lib/utils';
import { normalizeGraphicProjectByAPI } from './utils';
import { graphicElasticMock } from './graphicElasticMock';
import { graphicGraphqlMock } from './graphicGraphqlMock';

jest.mock( 'next/config', () => () => ( { publicRuntimeConfig: {} } ) );

jest.mock(
  'context/authContext',
  () => ( {
    useAuth: jest.fn( () => ( { user: { firstName: 'user' } } ) ),
  } ),
);

jest.mock( 'lib/browser', () => ( { updateUrl: jest.fn() } ) );

jest.mock(
  'next/router',
  () => ( {
    replace: jest.fn(),
    useRouter: jest.fn( () => ( {
      query: {
        content: 'graphic',
        id: 'ck8en7r8x0b7007652jpf9a59',
        site: 'commons.america.gov',
        language: 'en-us',
      },
    } ) ),
  } ),
);

jest.mock(
  'components/Notification/Notification',
  () => function Notification() { return ''; },
);
jest.mock(
  'components/Share/Share',
  () => function Share() { return ''; },
);
jest.mock(
  'components/TabLayout/TabLayout',
  () => function TabLayout() { return ''; },
);
jest.mock(
  'components/download/DownloadItem/DownloadItem',
  () => function DownloadItem() { return ''; },
);
jest.mock(
  './Download/GraphicFiles',
  () => function GraphicFiles() { return ''; },
);
jest.mock(
  './Download/GenericFiles',
  () => function GenericFiles() { return ''; },
);
jest.mock(
  './Download/Help',
  () => function Help() { return ''; },
);
jest.mock(
  'components/modals/ModalLangDropdown/ModalLangDropdown',
  () => function ModalLangDropdown() { return ''; },
);
jest.mock(
  '../modals/ModalImage/ModalImage',
  () => function ModalImage() { return ''; },
);
jest.mock(
  '../modals/ModalContentMeta/ModalContentMeta',
  () => function ModalContentMeta() { return ''; },
);
jest.mock(
  'components/modals/ModalDescription/ModalDescription',
  () => function ModalDescription() { return ''; },
);
jest.mock(
  'components/modals/ModalPostMeta/ModalPostMeta',
  () => function ModalPostMeta() { return ''; },
);
jest.mock(
  'components/modals/ModalPostTags/ModalPostTags',
  () => function ModalPostTags() { return ''; },
);

describe( '<GraphicProject />, for GraphQL data', () => {
  const props = {
    displayAsModal: true,
    isAdminPreview: true,
    item: graphicGraphqlMock,
    useGraphQl: true,
  };

  const normalizedData = normalizeGraphicProjectByAPI( {
    file: props.item,
    useGraphQl: props.useGraphQl,
  } );

  let Component;
  let wrapper;
  const selectedUnit = normalizedData.images.find( img => img.language.locale === 'en-us' );

  /* eslint-disable global-require */
  const { updateUrl } = require( 'lib/browser' );

  beforeEach( () => {
    Component = <GraphicProject { ...props } />;
    wrapper = mount( Component );
  } );

  it( 'renders without crashing', () => {
    expect( wrapper.exists() ).toEqual( true );
  } );

  it( 'renders the ModalItem with correct props', () => {
    const modalItem = wrapper.find( 'ModalItem' );
    const { title, language } = selectedUnit;
    const className = 'graphic-project adminPreview';

    expect( modalItem.exists() ).toEqual( true );
    expect( modalItem.prop( 'className' ) ).toEqual( className );
    expect( modalItem.prop( 'headline' ) ).toEqual( title );
    expect( modalItem.prop( 'lang' ) ).toEqual( language.language_code );
    expect( modalItem.prop( 'textDirection' ) )
      .toEqual( language.text_direction );
    expect( modalItem.prop( 'children' ).length ).toEqual( 9 );
  } );

  it( 'renders the preview Notification', () => {
    const notification = wrapper.find( 'Notification' );
    const notificationProps = {
      el: 'p',
      show: true,
      customStyles: {
        position: 'absolute',
        top: '0',
        left: '0',
        right: '0',
        padding: '1em 1.5em',
        fontSize: '1em',
        backgroundColor: '#fdb81e',
      },
      msg: 'This is a preview of your graphics project on Content Commons.',
    };

    expect( notification.exists() ).toEqual( true );
    expect( notification.props() ).toEqual( notificationProps );
  } );

  it( 'renders the ModalLanguageDropdown', () => {
    const langDropdown = wrapper.find( '.modal_options_left > ModalLangDropdown' );

    expect( langDropdown.exists() ).toEqual( true );
    expect( langDropdown.prop( 'item' ) ).toEqual( normalizedData );
    expect( langDropdown.prop( 'selected' ) )
      .toEqual( selectedUnit.language.display_name );
    expect( langDropdown.prop( 'handleLanguageChange' ).name ).toEqual( 'handleLanguageChange' );
  } );

  it( 'calling handleLanguageChange selects a new language unit', () => {
    const langDropdown = () => wrapper.find( '.modal_options_left > ModalLangDropdown' );
    const newLangSelection = 'French';

    act( () => {
      langDropdown().prop( 'handleLanguageChange' )( newLangSelection );
    } );
    wrapper.update();

    const newSelectedUnit = normalizedData.images.find( img => img.language.display_name === newLangSelection );
    const modalHeadline = wrapper.find( '.modal_headline' );

    expect( langDropdown().prop( 'selected' ) )
      .toEqual( newSelectedUnit.language.display_name );
    expect( modalHeadline.text() ).toEqual( newSelectedUnit.title );
    expect( updateUrl ).not.toHaveBeenCalled();
  } );

  it( 'renders the trigger container and its child components', () => {
    const triggerContainer = wrapper.find( '.trigger-container' );
    // use snapshot since child components are nested so deeply.

    expect( toJSON( triggerContainer ) ).toMatchSnapshot();
  } );

  it( 'renders the correct download tabs', () => {
    const { id } = props.item;
    const triggerContainer = wrapper.find( '.trigger-container' );
    const popover = triggerContainer.find( `[id="${id}_graphic-download"]` );
    const tabLayout = mount( popover.prop( 'children' ) );
    const { tabs } = tabLayout.props();
    const tabTitles = [
      'Graphic Files', 'Editable Files', 'Other', 'Help',
    ];

    expect( tabs.length ).toEqual( tabTitles.length );
    tabs.forEach( ( tab, i ) => {
      expect( tab.title ).toEqual( tabTitles[i] );
    } );
  } );

  it( 'renders the ModalImage thumbnail', () => {
    const modalImage = wrapper.find( 'ModalImage' );

    expect( modalImage.exists() ).toEqual( true );
    expect( modalImage.props() ).toEqual( {
      thumbnail: selectedUnit.srcUrl,
      thumbnailMeta: { alt: selectedUnit.alt },
    } );
  } );

  it( 'renders the ModalContentMeta', () => {
    const contentMeta = wrapper.find( 'ModalContentMeta' );
    const { modified, projectType } = normalizedData;

    expect( contentMeta.exists() ).toEqual( true );
    expect( contentMeta.props() ).toEqual( {
      type: `${projectType.toLowerCase().replace( '_', ' ' )} graphic`,
      dateUpdated: modified,
    } );
  } );

  it( 'renders the ModalDescription', () => {
    const contentDesc = wrapper.find( 'ModalDescription' );

    expect( contentDesc.exists() ).toEqual( true );
    expect( contentDesc.props() ).toEqual( {
      description: normalizedData.desc,
    } );
  } );

  it( 'renders the internal description', () => {
    const internalDesc = wrapper.find( '.graphic-project__content.internal-desc' );
    const title = internalDesc.find( '.graphic-project__content__title' );

    expect( internalDesc.exists() ).toEqual( true );
    expect( title.exists() ).toEqual( true );
    expect( title.text() ).toEqual( 'Internal Description:' );
    expect( internalDesc.contains( normalizedData.descInternal ) ).toEqual( true );
  } );

  it( 'renders the alt text', () => {
    const alt = wrapper.find( '.graphic-project__content.alt' );
    const title = alt.find( '.graphic-project__content__title' );

    expect( alt.exists() ).toEqual( true );
    expect( title.exists() ).toEqual( true );
    expect( title.text() ).toEqual( 'Alt (Alternative) Text:' );
    expect( alt.contains( selectedUnit.alt ) ).toEqual( true );
  } );

  it( 'renders the ModalPostMeta', () => {
    const postMeta = wrapper.find( 'ModalPostMeta' );
    const { owner, projectType, published } = normalizedData;

    expect( postMeta.exists() ).toEqual( true );
    expect( postMeta.props() ).toEqual( {
      type: projectType,
      logo: 'image-stub',
      source: owner,
      datePublished: published,
    } );
  } );

  it( 'renders the ModalPostTags', () => {
    const postTags = wrapper.find( 'ModalPostTags' );

    expect( postTags.exists() ).toEqual( true );
    expect( postTags.props() ).toEqual( {
      tags: getTransformedLangTaxArray( props.item.categories ),
    } );
  } );
} );

describe( '<GraphicProject />, for GraphQL data with no support files', () => {
  const noSupportFiles = {
    ...graphicGraphqlMock,
    supportFiles: [],
  };

  const props = {
    displayAsModal: true,
    isAdminPreview: true,
    item: noSupportFiles,
    useGraphQl: true,
  };

  let Component;
  let wrapper;

  beforeEach( () => {
    Component = <GraphicProject { ...props } />;
    wrapper = mount( Component );
  } );

  it( 'renders without crashing', () => {
    expect( wrapper.exists() ).toEqual( true );
  } );
} );

describe( '<GraphicProject />, for GraphQL data with null support files', () => {
  const noSupportFiles = {
    ...graphicGraphqlMock,
    supportFiles: null,
  };

  const props = {
    displayAsModal: true,
    isAdminPreview: true,
    item: noSupportFiles,
    useGraphQl: true,
  };

  let Component;
  let wrapper;

  beforeEach( () => {
    Component = <GraphicProject { ...props } />;
    wrapper = mount( Component );
  } );

  it( 'renders without crashing', () => {
    expect( wrapper.exists() ).toEqual( true );
  } );
} );

describe( '<GraphicProject />, for ElasticSearch data', () => {
  const props = {
    item: graphicElasticMock[0]._source,
  };

  const normalizedData = normalizeGraphicProjectByAPI( {
    file: props.item,
  } );

  let Component;
  let wrapper;
  const selectedUnit = normalizedData.images.find( img => img.language.locale === 'en-us' );

  const { updateUrl } = require( 'lib/browser' );

  beforeEach( () => {
    Component = <GraphicProject { ...props } />;
    wrapper = mount( Component );
  } );

  it( 'renders without crashing', () => {
    expect( wrapper.exists() ).toEqual( true );
  } );

  it( 'renders the ModalItem with correct props', () => {
    const modalItem = wrapper.find( 'ModalItem' );
    const { title, language } = selectedUnit;

    expect( modalItem.exists() ).toEqual( true );
    expect( modalItem.prop( 'className' ) ).toEqual( 'graphic-project' );
    expect( modalItem.prop( 'headline' ) ).toEqual( title );
    expect( modalItem.prop( 'lang' ) ).toEqual( language.language_code );
    expect( modalItem.prop( 'textDirection' ) )
      .toEqual( language.text_direction );
    expect( modalItem.prop( 'children' ).length ).toEqual( 9 );
  } );

  it( 'does not render the preview Notification', () => {
    const notification = wrapper.find( 'Notification' );

    expect( notification.exists() ).toEqual( false );
  } );

  it( 'renders the ModalLanguageDropdown', () => {
    const langDropdown = wrapper.find( '.modal_options_left > ModalLangDropdown' );

    expect( langDropdown.exists() ).toEqual( true );
    expect( langDropdown.prop( 'item' ) ).toEqual( normalizedData );
    expect( langDropdown.prop( 'selected' ) )
      .toEqual( selectedUnit.language.display_name );
    expect( langDropdown.prop( 'handleLanguageChange' ).name ).toEqual( 'handleLanguageChange' );
  } );

  it( 'calling handleLanguageChange selects a new language unit', () => {
    const langDropdown = () => wrapper.find( '.modal_options_left > ModalLangDropdown' );
    const newLangSelection = 'French';

    act( () => {
      langDropdown().prop( 'handleLanguageChange' )( newLangSelection );
    } );
    wrapper.update();

    const newSelectedUnit = normalizedData.images.find( img => img.language.display_name === newLangSelection );
    const modalHeadline = wrapper.find( '.modal_headline' );
    const { id, site, type } = props.item;
    const { language, title } = newSelectedUnit;
    const newUrl = `/${type}?id=${id}&site=${site}&language=${language.locale}`;

    expect( langDropdown().prop( 'selected' ) )
      .toEqual( language.display_name );
    expect( modalHeadline.text() ).toEqual( title );
    expect( updateUrl ).toHaveBeenCalledWith( newUrl );
  } );

  it( 'renders the trigger container and its child components', () => {
    const triggerContainer = wrapper.find( '.trigger-container' );
    // use snapshot since child components are nested so deeply.

    expect( toJSON( triggerContainer ) ).toMatchSnapshot();
  } );

  it( 'renders the correct download tabs', () => {
    const { id } = props.item;
    const triggerContainer = wrapper.find( '.trigger-container' );
    const popover = triggerContainer.find( `[id="${id}_graphic-download"]` );
    const tabLayout = mount( popover.prop( 'children' ) );
    const { tabs } = tabLayout.props();
    const tabTitles = [
      'Graphic Files', 'Editable Files', 'Other', 'Help',
    ];

    expect( tabs.length ).toEqual( tabTitles.length );
    tabs.forEach( ( tab, i ) => {
      expect( tab.title ).toEqual( tabTitles[i] );
    } );
  } );

  it( 'renders the ModalImage thumbnail', () => {
    const modalImage = wrapper.find( 'ModalImage' );

    expect( modalImage.exists() ).toEqual( true );
    expect( modalImage.props() ).toEqual( {
      thumbnail: selectedUnit.srcUrl,
      thumbnailMeta: { alt: selectedUnit.alt },
    } );
  } );

  it( 'renders the ModalContentMeta', () => {
    const contentMeta = wrapper.find( 'ModalContentMeta' );
    const { modified, projectType } = normalizedData;

    expect( contentMeta.exists() ).toEqual( true );
    expect( contentMeta.props() ).toEqual( {
      type: `${projectType.toLowerCase().replace( '_', ' ' )} graphic`,
      dateUpdated: modified,
    } );
  } );

  it( 'renders the ModalDescription', () => {
    const contentDesc = wrapper.find( 'ModalDescription' );

    expect( contentDesc.exists() ).toEqual( true );
    expect( contentDesc.props() ).toEqual( {
      description: normalizedData.desc,
    } );
  } );

  it( 'renders the internal description', () => {
    const internalDesc = wrapper.find( '.graphic-project__content.internal-desc' );
    const title = internalDesc.find( '.graphic-project__content__title' );

    expect( internalDesc.exists() ).toEqual( true );
    expect( title.exists() ).toEqual( true );
    expect( title.text() ).toEqual( 'Internal Description:' );
    expect( internalDesc.contains( normalizedData.descInternal ) ).toEqual( true );
  } );

  it( 'renders the alt text', () => {
    const alt = wrapper.find( '.graphic-project__content.alt' );
    const title = alt.find( '.graphic-project__content__title' );

    expect( alt.exists() ).toEqual( true );
    expect( title.exists() ).toEqual( true );
    expect( title.text() ).toEqual( 'Alt (Alternative) Text:' );
    expect( alt.contains( selectedUnit.alt ) ).toEqual( true );
  } );

  it( 'renders the ModalPostMeta', () => {
    const postMeta = wrapper.find( 'ModalPostMeta' );
    const { owner, projectType, published } = normalizedData;

    expect( postMeta.exists() ).toEqual( true );
    expect( postMeta.props() ).toEqual( {
      type: projectType,
      logo: 'image-stub',
      source: owner,
      datePublished: published,
    } );
  } );

  it( 'renders the ModalPostTags', () => {
    const postTags = wrapper.find( 'ModalPostTags' );

    expect( postTags.exists() ).toEqual( true );
    expect( postTags.props() ).toEqual( {
      tags: props.item.categories,
    } );
  } );
} );

describe( '<GraphicProject />, for ElasticSearch data with no support files', () => {
  const noSupportFiles = {
    ...graphicElasticMock[0]._source,
    supportFiles: [],
  };

  const props = {
    item: noSupportFiles,
  };

  let Component;
  let wrapper;

  beforeEach( () => {
    Component = <GraphicProject { ...props } />;
    wrapper = mount( Component );
  } );

  it( 'renders without crashing', () => {
    expect( wrapper.exists() ).toEqual( true );
  } );
} );

describe( '<GraphicProject />, for ElasticSearch data with null support files', () => {
  const noSupportFiles = {
    ...graphicElasticMock[0]._source,
    supportFiles: null,
  };

  const props = {
    item: noSupportFiles,
  };

  let Component;
  let wrapper;

  beforeEach( () => {
    Component = <GraphicProject { ...props } />;
    wrapper = mount( Component );
  } );

  it( 'renders without crashing', () => {
    expect( wrapper.exists() ).toEqual( true );
  } );
} );
