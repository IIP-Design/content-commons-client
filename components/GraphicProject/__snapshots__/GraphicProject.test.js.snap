// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GraphicProject />, for ElasticSearch data renders the trigger container and its child components 1`] = `
<div
  className="trigger-container"
>
  <Popover
    className="graphic-project__popover graphic-project__popover--share"
    expandFromRight={true}
    id="ck8h90w6x030l0765onryiqdt_graphic-share"
    toolTip="Share graphic"
    trigger={
      <img
        alt="share icon"
        src="image-stub"
        style={
          Object {
            "height": "20px",
            "width": "20px",
          }
        }
      />
    }
  >
    <span
      className="popover graphic-project__popover graphic-project__popover--share"
    >
      <button
        aria-controls="popoverContent"
        aria-expanded={false}
        aria-haspopup="true"
        className="popover_trigger"
        id="popoverButton_ck8h90w6x030l0765onryiqdt_graphic-share"
        onClick={[Function]}
        tooltip="Share graphic"
        type="button"
      >
        <img
          alt="share icon"
          src="image-stub"
          style={
            Object {
              "height": "20px",
              "width": "20px",
            }
          }
        />
      </button>
    </span>
  </Popover>
  <Popover
    className="graphic-project__popover graphic-project__popover--download"
    expandFromRight={true}
    id="ck8h90w6x030l0765onryiqdt_graphic-download"
    toolTip="Download graphic"
    trigger={
      <img
        alt="download icon"
        src="image-stub"
        style={
          Object {
            "height": "20px",
            "width": "20px",
          }
        }
      />
    }
  >
    <span
      className="popover graphic-project__popover graphic-project__popover--download"
    >
      <button
        aria-controls="popoverContent"
        aria-expanded={false}
        aria-haspopup="true"
        className="popover_trigger"
        id="popoverButton_ck8h90w6x030l0765onryiqdt_graphic-download"
        onClick={[Function]}
        tooltip="Download graphic"
        type="button"
      >
        <img
          alt="download icon"
          src="image-stub"
          style={
            Object {
              "height": "20px",
              "width": "20px",
            }
          }
        />
      </button>
    </span>
  </Popover>
</div>
`;

exports[`<GraphicProject />, for GraphQL data renders the trigger container and its child components 1`] = `
<div
  className="trigger-container"
>
  <Popover
    className="graphic-project__popover graphic-project__popover--share"
    expandFromRight={true}
    id="ck8en7r8x0b7007652jpf9a59_graphic-share"
    toolTip="Share graphic"
    trigger={
      <img
        alt="share icon"
        src="image-stub"
        style={
          Object {
            "height": "20px",
            "width": "20px",
          }
        }
      />
    }
  >
    <span
      className="popover graphic-project__popover graphic-project__popover--share"
    >
      <button
        aria-controls="popoverContent"
        aria-expanded={false}
        aria-haspopup="true"
        className="popover_trigger"
        id="popoverButton_ck8en7r8x0b7007652jpf9a59_graphic-share"
        onClick={[Function]}
        tooltip="Share graphic"
        type="button"
      >
        <img
          alt="share icon"
          src="image-stub"
          style={
            Object {
              "height": "20px",
              "width": "20px",
            }
          }
        />
      </button>
    </span>
  </Popover>
  <Popover
    className="graphic-project__popover graphic-project__popover--download"
    expandFromRight={true}
    id="ck8en7r8x0b7007652jpf9a59_graphic-download"
    toolTip="Download graphic"
    trigger={
      <img
        alt="download icon"
        src="image-stub"
        style={
          Object {
            "height": "20px",
            "width": "20px",
          }
        }
      />
    }
  >
    <span
      className="popover graphic-project__popover graphic-project__popover--download"
    >
      <button
        aria-controls="popoverContent"
        aria-expanded={false}
        aria-haspopup="true"
        className="popover_trigger"
        id="popoverButton_ck8en7r8x0b7007652jpf9a59_graphic-download"
        onClick={[Function]}
        tooltip="Download graphic"
        type="button"
      >
        <img
          alt="download icon"
          src="image-stub"
          style={
            Object {
              "height": "20px",
              "width": "20px",
            }
          }
        />
      </button>
    </span>
  </Popover>
</div>
`;
